### PR TITLE
feat: Attach multiple files in the MessageInput

### PR DESCRIPTION
**Pull request by HoonBaek**
[UIKIT-4405](https://sendbird.atlassian.net/browse/UIKIT-4405)

### Feat
* Create `useHandleUploadFiles` hook for handling each case of sending file&files messages
  * Condition & Behavior
    1. If selected only one file -> send file message
    2. If selected multiple files
       * If selected only images -> send MFM
       * If selected several type of files -> send MFM for Images & send file message for the other files

[UIKIT-4405]: https://sendbird.atlassian.net/browse/UIKIT-4405?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: scripts/index_d_ts [hunk 1]
```diff
@@ -2258,7 +2258,7 @@ declare module '@sendbird/uikit-react/ui/MessageInput' {
     name: string,
     placeholder?: string,
     maxLength?: number,
-    onFileUpload?: () => void,
+    onFileUpload?: (files: Array<File>) => void,
```
Comment (HoonBaek) [inline scripts/index_d_ts:2261]: let me make another PR for this job

File: src/lib/types.ts [hunk 2]
```diff
@@ -88,6 +89,7 @@ export interface SendBirdStateConfig {
   isReactionEnabled: boolean;
   isMentionEnabled: boolean;
   isVoiceMessageEnabled?: boolean;
+  uikitMultipleFilesMessageLimit: Nullable<number>; // shouldn't be nullable
```
Comment (HoonBaek) [inline src/lib/types.ts:92]: We have to set a default value, isn't it? @liamcho

File: src/modules/Channel/context/ChannelProvider.tsx [hunk 3]
```diff
@@ -421,6 +424,15 @@ const ChannelProvider: React.FC<ChannelContextProps> = (props: ChannelContextPro
       messagesDispatcher,
     },
   );
+  const [sendMultipleFilesMessage] = useSendMultipleFilesMessage({
+    currentChannel: currentGroupChannel,
+    // Open interface to <Channel /> & <ChannelProvider >
+    // onBeforeSendMultipleFilesMessage: (params) => params,
```
Comment (HoonBaek) [inline src/modules/Channel/context/ChannelProvider.tsx:430]: This too, let me make a different PR and update types

File: src/lib/types.ts (hunk 2 above)
Comment (liamcho) [reply src/lib/types.ts:92]: ah yes you are right. Thanks for the catch. Let me create fix pr for this.
Screen Shot 2023-09-11 at 2 55 13 PM

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  const uikitMultipleFilesMessageLimit = config?.uikitMultipleFilesMessageLimit ?? 15;
+
+  const handleUploadeFiles = useCallback((fileList: FileList) => {
+    const files: File[] = Array.from(fileList);
+
+    if (!sendFileMessage || !sendMultipleFilesMessage) {
+      logger.warning('Channel|useHandleUploadFiles: required functions are undefined', { sendFileMessage, sendMultipleFilesMessage });
+      return;
+    }
+    if (files.length > 0) {
```
Comment (AhyoungRyu) [inline src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:41]: Small suggestion; using ts-pattern in here would be better in terms of readability. 
```
import { match } from 'ts-pattern'
match(files.length)
  // put any child if-else inside of when stmt 
  .when(n => n  > 0, () => {...})
  .when(n => n  === 1, () => {...})
  .when(n => n  > 1, () => {...})
```

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        if (isImage(file.type)) {
+          imageFiles.push(file);
+        } else {
+          otherFiles.push(file);
+        }
+      });
+
+      if (otherFiles.length > 0) {
+        otherFiles.forEach((file: File) => {
+          sendFileMessage(file, quoteMessage);
```
Comment (AhyoungRyu) [inline src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:68]: Don't we need an error handling here(same for the below `sendMultipleFilesMessage`)? Guess it returns promise and looks there's possibility we get an error here.

File: src/ui/MessageInput/index.jsx [hunk 6]
```diff
@@ -523,7 +518,13 @@ const MessageInput = React.forwardRef((props, ref) => {
                   className="sendbird-message-input--attach-input"
                   type="file"
                   ref={fileInputRef}
-                  onChange={handleUploadFile(onFileUpload)}
+                  // It will affect to <Channel /> and <Thread />
+                  onChange={(event) => {
+                    onFileUpload([...event.currentTarget.files]);
```
Comment (AhyoungRyu) [inline src/ui/MessageInput/index.jsx:523]: Do we need to copy the original array of files and make a new files array here? If not, this would just work.

```suggestion
                    onFileUpload(event.currentTarget.files);
```

File: src/lib/types.ts (hunk 2 above)
Comment (liamcho) [reply src/lib/types.ts:92]: https://github.com/sendbird/sendbird-uikit-react/pull/728

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+export const useHandleUploadFiles = ({
+  sendFileMessage,
+  sendMultipleFilesMessage,
+  quoteMessage,
+}: UseHandleUploadFilesDynamicProps, {
+  logger,
+}: UseHandleUploadFilesStaticProps): Array<HandleUploadFunctionType> => {
+  const { config } = useSendbirdStateContext();
+  const uikitMultipleFilesMessageLimit = config?.uikitMultipleFilesMessageLimit ?? 15;
```
Comment (liamcho) [inline src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:32]: no need for ?? 15?
[This PR](https://github.com/sendbird/sendbird-uikit-react/pull/728/files) should set default to 10.

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+  const handleUploadeFiles = useCallback((fileList: FileList) => {
+    const files: File[] = Array.from(fileList);
+
+    if (!sendFileMessage || !sendMultipleFilesMessage) {
+      logger.warning('Channel|useHandleUploadFiles: required functions are undefined', { sendFileMessage, sendMultipleFilesMessage });
+      return;
+    }
+    if (files.length > 0) {
+      if (files.length > uikitMultipleFilesMessageLimit) {
```
Comment (liamcho) [inline src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:42]: should be `>=` ?

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts (hunk 4 above)
Comment (liamcho) [inline src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:41]: Maybe remove this and add
```
if (files.length === 0) return; 
```
at the top instead?

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        logger.info(`Channel|useHandleUploadeFiles: Cannot upload files more than ${uikitMultipleFilesMessageLimit}`);
+        return;
+      }
+      if (files.some((file: File) => file.size > FILE_SIZE_LIMIT)) {
+        logger.info(`Channel|useHandleUploadeFiles: Cannot upload file size exceeding ${FILE_SIZE_LIMIT}`);
+        return;
+      }
+    }
+
+    if (files.length === 1) {
```
Comment (liamcho) [inline src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:52]: How about organize like below?
```
if (files.legnth === 0) return;
if (files.length === 1) {
  // sendFileMessage
} else {
  // 1. check uikitMultipleFilesMessageLimit
  // 2. sendFileMessage(s) and/or sendMultipleFilesMessage
}
```

File: src/modules/Channel/context/ChannelProvider.tsx (hunk 3 above)
Comment (liamcho) [reply src/modules/Channel/context/ChannelProvider.tsx:430]: Also, we have to give prompt to user if files limit / size validation fails.

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+      if (otherFiles.length > 0) {
+        otherFiles.forEach((file: File) => {
+          sendFileMessage(file, quoteMessage);
+        });
+      }
+      if (imageFiles.length === 1) {
+        sendFileMessage(imageFiles[0], quoteMessage);
+      } else if (imageFiles.length > 1) {
+        sendMultipleFilesMessage(imageFiles, quoteMessage);
```
Comment (liamcho) [inline src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:74]: We were suppose to send MFMs before sending other file messages right? I can't find any note on this.
https://docs.google.com/document/d/1Ww1rixAuL0fK6TTBOhWcBn5QB4jCJoDH1OSoi5DyfUY/edit

File: src/modules/Channel/context/ChannelProvider.tsx (hunk 3 above)
Comment (HoonBaek) [reply src/modules/Channel/context/ChannelProvider.tsx:430]: Yes, it will be included in the other PR

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts (hunk 7 above)
Comment (HoonBaek) [reply src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:32]: oh, this was for my self-testing, let me fix it!

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts (hunk 8 above)
Comment (liamcho) [reply src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:42]: oh my bad its no problem

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts (hunk 9 above)
Comment (liamcho) [reply src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:52]: lets use match instead.

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts (hunk 4 above)
Comment (HoonBaek) [reply src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:41]: that's a good idea!

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts (hunk 8 above)
Comment (HoonBaek) [reply src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:42]: Do you mean if a user selects 10 files, we should show the notice?

File: src/ui/MessageInput/index.jsx (hunk 6 above)
Comment (HoonBaek) [reply src/ui/MessageInput/index.jsx:523]: Oh, this is necessary I think.
The `event.currentTarget` object becomes empty soon because I reset the value right below.
```
event.target.value = '';
```

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts (hunk 5 above)
Comment (HoonBaek) [reply src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:68]: The actual sending logic is inside of each useCallback hook.
the sendFileMessage is all right, however, sendMFM doesn't have error-handling logic.

so let me add it to the useSendMultipleFilesMessage hook! thank you for the good catch!

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts (hunk 10 above)
Comment (HoonBaek) [reply src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:74]: image
[Yes, here it is!](https://sendbird.atlassian.net/wiki/spaces/UK/pages/2118778993/Multiple+files+message)

File: scripts/index_d_ts (hunk 1 above)
Comment (HoonBaek) [reply scripts/index_d_ts]: https://github.com/sendbird/sendbird-uikit-react/pull/730

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+  sendFileMessage,
+  sendMultipleFilesMessage,
+  quoteMessage,
+}: UseHandleUploadFilesDynamicProps, {
+  logger,
+}: UseHandleUploadFilesStaticProps): Array<HandleUploadFunctionType> => {
+  const { config } = useSendbirdStateContext();
+  const uikitMultipleFilesMessageLimit = config?.uikitMultipleFilesMessageLimit;
+
+  const handleUploadeFiles = useCallback((fileList: FileList) => {
```
Comment (liamcho) [inline src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:36]: minor: typo.

handleUploadeFiles X
handleUploadFiles O

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    if (!sendFileMessage || !sendMultipleFilesMessage) {
+      logger.warning('Channel|useHandleUploadFiles: required functions are undefined', { sendFileMessage, sendMultipleFilesMessage });
+      return;
+    }
+    if (files.length === 0) {
+      logger.warning('Channel|usehandleUploadFiles: given file list is empty.', { files });
+      return;
+    }
+    if (files.length > uikitMultipleFilesMessageLimit) {
+      logger.info(`Channel|useHandleUploadeFiles: Cannot upload files more than ${uikitMultipleFilesMessageLimit}`);
```
Comment (liamcho) [inline src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:49]:
```suggestion
      logger.info(`Channel|useHandleUploadFiles: Cannot upload files more than ${uikitMultipleFilesMessageLimit}`);
```

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+      return;
+    }
+    if (files.some((file: File) => file.size > FILE_SIZE_LIMIT)) {
+      logger.info(`Channel|useHandleUploadeFiles: Cannot upload file size exceeding ${FILE_SIZE_LIMIT}`);
```
Comment (liamcho) [inline src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:53]:
```suggestion
      logger.info(`Channel|useHandleUploadFiles: Cannot upload file size exceeding ${FILE_SIZE_LIMIT}`);
```

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts (hunk 11 above)
Comment (liamcho) [inline src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:36]:
```suggestion
  const handleUploadFiles = useCallback((fileList: FileList) => {
```

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+          });
+        }
+      });
+  }, [
+    sendFileMessage,
+    sendMultipleFilesMessage,
+    quoteMessage,
+  ]);
+
+  return [handleUploadeFiles];
```
Comment (liamcho) [inline src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:90]:
```suggestion
  return [handleUploadFiles];
```

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+/**
+ * The handleUploadFiles is a function sending a FileMessage and MultipleFilesMessage
+ * by the received FileList from the ChangeEvent of MessageInput component.
+ */
+
+const FILE_SIZE_LIMIT = 300000000; // 300MB
+
+interface UseHandleUploadFilesDynamicProps {
+  sendFileMessage: (file: File, quoteMessage?: SendableMessageType) => void;
```
Comment (liamcho) [inline src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:17]: how about replace with
`SendFileMessageFunctionType` ?

**Review by liamcho — COMMENTED** [review]
BTW, we should fix this in ChannelProvider.tsx
Screen Shot 2023-09-12 at 12 07 48 PM

what is diff btw
- useSendFileMessage
- useSendFileMessageCallback
?

It seems callback is used here.

File: src/modules/Channel/context/ChannelProvider.tsx [hunk 16]
```diff
@@ -152,8 +154,9 @@ interface ChannelProviderInterface extends ChannelContextProps, MessageStoreInte
   resendMessage(failedMessage: SendableMessageType): Promise<SendableMessageType>,
   // TODO: Good to change interface to using params / This part need refactoring
   sendMessage(props: SendMessageParams): Promise<UserMessage>,
-  sendFileMessage(file: File, quoteMessage: SendableMessageType): Promise<FileMessage>,
+  sendFileMessage(file: File, quoteMessage?: SendableMessageType): Promise<FileMessage>,
   sendVoiceMessage: (file: File, duration: number, quoteMessage?: SendableMessageType) => void,
+  sendMultipleFilesMessage: (files: Array<File>, quoteMessage?: SendableMessageType) => Promise<MultipleFilesMessage>,
```
Comment (liamcho) [inline src/modules/Channel/context/ChannelProvider.tsx:159]: return type is different than `useSendMultipleFilesMessage()`. It's void there.

**Review by liamcho — COMMENTED** [review]
no sendMultipleFilesMessage in `ThreadProvider.tsx`?

File: src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts (hunk 15 above)
Comment (HoonBaek) [reply src/modules/Channel/components/MessageInput/useHandleUploadFiles.ts:17]: good idea!

File: src/modules/Channel/context/ChannelProvider.tsx (hunk 16 above)
Comment (HoonBaek) [reply src/modules/Channel/context/ChannelProvider.tsx:159]: Oh, it's a great catch! It seems the other `sendMessage` and `sendFileMessage` also have kept wrong return types!

**Comment by HoonBaek** [comment]
@liamcho 
`useSendFileMessage` and `useSendFileMessageCallback` are almost the same, they are the custom useCallback hooks that send file messages under the Channel and Thread modules.

**Review by liamcho — APPROVED** [review]
LGTM! jsut make sure to check if ThreadProvider.tsx needs to be supporting MFM too!

**Comment by HoonBaek** [comment]
I make it possible to send MFM from Thread also!
